packages/core/utils/src/serializer.js

```diff
@@ -16,37 +17,160 @@ export function registerSerializableClass(name: string, ctor: Class<*>) {
-        return {
-          $$type: type,
-          value: {...serialized}
-        };
+    if (object instanceof Map) {
+      return new Map(object);
+    }
+
+    if (object instanceof Set) {
+      return new Set(object);
+    }
+
+    return Object.create(
+      Object.getPrototypeOf(object),
+      Object.getOwnPropertyDescriptors(object)
+    );
+  }
+
+  return object;
+}
+
+function mapObject(object: any, fn: (val: any) => any, preOrder = false) {
+  let cache = new Map();
+  let memo = new Map();
+
+  // Memoize the passed function to ensure it always returns the exact same
+  // output by reference for the same input. This is important to maintain
+  // reference integrity when deserializing rather than cloning.
+  let memoizedFn = (val: any) => {
+    if (memo.has(val)) {
+      return memo.get(val);
+    }
+
+    let res = fn(val);
+    memo.set(val, res);
+    return res;
+  };
+
+  let walk = (object: any, shouldCopy = false) => {
+    // Check the cache first, both for performance and cycle detection.
+    if (cache.has(object)) {
+      return cache.get(object);
+    }
+
+    let result = object;
+    cache.set(object, result);
+
+    let processKey = (key: any, value: any) => {
+      let newValue = value;
+      if (preOrder) {
+        newValue = memoizedFn(value);
+      }
+
+      // Recursively walk the children
+      if (newValue && typeof newValue === 'object') {
+        newValue = walk(newValue, newValue === value);
+      }
+
+      if (!preOrder) {
+        newValue = memoizedFn(newValue);
+      }
+
+      if (newValue !== value) {
+        // Copy on write. We only need to do this when serializing, not deserializing.
+        if (object === result && preOrder && shouldCopy) {
+          result = shallowCopy(object);
+          cache.set(object, result);
+        }
+
+        // Replace the key with the new value
+        if (result instanceof Map) {
+          result.set(key, newValue);
+        } else if (result instanceof Set) {
+          // TODO: do we care about iteration order??
+          result.delete(value);
+          result.add(newValue);
+        } else {
+          result[key] = newValue;
+        }
+      }
+    };
+
+    // Iterate in various ways depending on type.
+    if (Array.isArray(object)) {
+      for (let i = 0; i < object.length; i++) {
+        processKey(i, object[i]);
+      }
+    } else if (object instanceof Map || object instanceof Set) {
+      for (let [key, val] of object.entries()) {
+        processKey(key, val);
+      }
+    } else {
+      for (let key in object) {
+        processKey(key, object[key]);
       }
     }
 
-    return serialized;
-  });
+    return result;
+  };
+
+  let mapped = memoizedFn(object);
+  return walk(mapped, mapped === object);
+}
+
+export function serialize(object: any): Buffer {
+  let mapped = mapObject(
+    object,
+    value => {
+      let serialized = value;
+      if (value && typeof value.serialize === 'function') {
```

Comment:
If this is ever used to serialize user-defined objects/classes, it would be safer to use a symbol e.g.:

```javascript
const SERIALIZE = Symbol()

// or (for use outside of Parcel)

const SERIALIZE = Symbol.for('parceljs.serialize')

// ...

if (value && typeof value[SERIALIZE] === 'function') { ... }
``` 
Even if the hook is only used internally, it might still better communicate the intention and avert future surprises.

## See Also

- https://www.keithcirkel.co.uk/metaprogramming-in-es6-symbols/#3-giving-developers-ability-to-add-hooks-to-their-objects-through-your-api
- https://nodejs.org/api/util.html#util_util_inspect_custom

Reply:
That's a good suggestion. We could also only allow a `serialize` method for constructors that are registered with the serializer.